### DISCUSSION TITLE
A small suggestion for an improvement.

### DISCUSSION THREAD

**Discussion by luhanzhen** [chocoteam/choco-solver#1029] [General]
Hi, i'm a leaner for constraint programming,  and when i read the source code, i found  a little problem.
for these two files in /choco/src/main/java/org/chocosolver/util/objects/setDataStructures/bitset:
your iterator for the bitset is here:
[	
```
public ISetIterator newIterator(){
		return new ISetIterator() {
			private int current = -1;
			@Override
			public void reset() {
				current = -1;
			}
			@Override
			public boolean hasNext() {
				return values.nextSetBit(current+1) >= 0;
			}
			@Override
			public int nextInt() {
				current = values.nextSetBit(current + 1);
				return current+offset;
			}
		};
	}
```
](https://github.com/chocoteam/choco-solver/tree/master/solver/src/main/java/org/chocosolver/util/objects/setDataStructures/bitset)

I think the method "values.nextSetBit(current+1)" is called twice when “hasNext()” and “nextInt()”  are in a for loop at the same time.
Maybe we can avoid this situation:

```
 public ISetIterator newIterator() {
        return new ISetIterator() {
            private int current = -1;
            private int next = -1;

            @Override
            public void reset() {
                current = -1;
                next = values.nextSetBit(0);
            }

            @Override
            public boolean hasNext() {
                next = values.nextSetBit(current + 1);
                return next >= 0;
            }

            @Override
            public int nextInt() {
                current = next;
                return current + offset;
            }
        };
    }
```
I have tested the above code and there seems to be no errors and an improvement in efficiency.
I'm looking forward to the authors' replies. 😊😊😊😊

**Comment by jgFages** [comment]
Hello,

I understand this can be seen as a double operation but it is necessary as hasNext() method should not modify the state of the object. Calling it twice should lead to the same result. 

If saving an operation is that critical for your application, maybe you can do your own loop relying on nextInt() only and analysing if the value returned is >= offset or not, but I would not recommend it.

**Comment by cprudhom** [comment]
> Calling it twice should lead to the same result.

In the proposed code, calling `hasNext()` twice leads to the same result. 
So, if it saves an operation, I think it can worth a try

**Comment by cprudhom** [comment]
BUT, it seems that the `next()` operation is sometimes called without calling `hasNext()` first, which makes the fix incorrect

**Comment by luhanzhen** [comment]
Perhaps I misunderstood, and now  I am learning the design principle of the iterators in Choco. 😄😄😄
